src/cargo/ops/registry/publish.rs

```diff
@@ -498,3 +564,152 @@ fn transmit(
 
     Ok(())
 }
+
+// State for tracking dependencies during upload.
+struct PublishOrder {
+    // The reverse-dependency graph. It has an edge p -> q if package q depends on package p.
+    rev_graph: Graph<PackageId, ()>,
+    // The weight of a package is the number of unpublished dependencies it has.
+    weights: HashMap<PackageId, usize>,
+}
+
+impl PublishOrder {
+    // Given a package dependency graph, creates a `PublishOrder` for tracking state and also
+    // a list of packages that have no dependencies (i.e. can be published immediately).
+    fn new(graph: &Graph<PackageId, ()>) -> (Self, Vec<PackageId>) {
```

Comment:
API wise this feels weird that we return `ready` as well as `Self`.

Options
- We keep it.  Its all internal
- We add a `take_ready` which removes items with `weight == 0`
- likely other approaches as well

Reply:
The API is definitely a little weird, but it has the advantage that it requires the least state-tracking: when you mark a package as finished, it's immediately obvious (in constant time, with no extra state) which packages are ready. To have a two-step API that separates out the "mark as done" and "get next ready" steps, we'd have to either add API for mutating `Graph` or (better, IMO) add an extra field to store packages that are ready to be taken in a `take_ready` function.

Given the extra complication, I prefer the "keep it" option...

Reply:
Do we need to mutate the graph or only `weights`?  If we did so, then wouldn't it still be trivial to track what packages are ready (its `self.weights.keys()`).

Reply:
Oh sorry, for some reason I misunderstood that you were talking about the `new` function. Yes, this can have a nicer API.